### PR TITLE
:bug: ensure correct shutdown handlers called

### PR DESCRIPTION
**Pull request by joerunde**
**What this PR does / why we need it**:

This was a tangent I went on today. This PR both:
- allows multiple servers to cleanly register shutdown handlers for interruption
- Updates the http server hackery to ensure that previously registered signal handlers are actually invoked when uvicorn overrides all the signal handlers

**Special notes for your reviewer**:

This is a bit of a departure from how we were explicitly setting signal handlers in `main` methods before, but this allows us to easily handle uvicorn's insistence on overriding all signal handlers.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: caikit/runtime/__main__.py [hunk 1]
```diff
@@ -1,6 +1,16 @@
-# Standard
-import signal
-
+# Copyright The Caikit Authors
```
Comment (gabe-l-hart) [inline caikit/runtime/__main__.py:1]: Whoops! Good catch

File: caikit/runtime/http_server/http_server.py [hunk 2]
```diff
@@ -282,11 +277,14 @@ def stop(self):
             grace_period_seconds (Union[float, int]): Grace period for service shutdown.
                 Defaults to application config
         """
-        self.server.should_exit = True
```
Comment (gabe-l-hart) [inline caikit/runtime/http_server/http_server.py:285]: Why was this moved? I think it makes sense to put it under the `if`, I'm just curious if having it outside was causing some specific problem

File: caikit/runtime/server_base.py [hunk 3]
```diff
@@ -137,6 +137,43 @@ def interrupt(self, signal_, _stack_frame):
...
+    def _add_signal_handler(sig, handler):
+        def nested_interrupt_builder(*handlers):
+            """Build and return an interrupt handler that calls all of *handlers"""
+
+            log.debug("Building interrupt handler: %s", handlers)
+
+            def interrupt(signal_, _stack_frame):
+                for handler in handlers:
+                    # Only call the handler if it is a callable fn that is _not_ a default handler
+                    log.debug("Running interrupt handler: %s", handler)
```
Comment (gabe-l-hart) [inline caikit/runtime/server_base.py:159]: Do we have to worry about writing to `stdout` in a signal handler? I know I've seen issues with writing to closed pipes before when logging in `__del__` functions

File: caikit/runtime/server_base.py [hunk 4]
```diff
@@ -137,6 +137,43 @@ def interrupt(self, signal_, _stack_frame):
...
+                        and handler is not signal.default_int_handler
+                    ):
+                        handler(signal_, _stack_frame)
+
+            return interrupt
+
+        try:
+            signal.signal(sig, nested_interrupt_builder(handler, signal.getsignal(sig)))
+        except ValueError:
+            log.warning(
```
Comment (gabe-l-hart) [inline caikit/runtime/server_base.py:173]: Do we expect this to indicate a problem? Seems like putting this on `warning` would likely cause confusion if this is a "working as expected" usecase when running either of the servers in a child thread.

**Review by gabe-l-hart — COMMENTED** [review]
A couple of questions about some of the logging. Also, I don't see any unit tests. I expect that this would be quite hard to actually test since it only works with the primary process, but we probably could do something like [this test](https://github.com/caikit/caikit/blob/main/tests/runtime/test_grpc_server.py#L1331) where we launch the server in a subprocess so that we can send it a signal

File: caikit/runtime/http_server/http_server.py (hunk 2 above)
Comment (joerunde) [reply caikit/runtime/http_server/http_server.py:285]: Nah I was just being a bit more specific- it's only needed to signal the server running in another thread to stop

File: caikit/runtime/server_base.py (hunk 4 above)
Comment (joerunde) [reply caikit/runtime/server_base.py:173]: true, I guess this should be info level

File: caikit/runtime/server_base.py (hunk 3 above)
Comment (joerunde) [reply caikit/runtime/server_base.py:159]: uhhh, well we were logging before so I kept logging here? But that's a good question.

FWIW the uvicorn handler definitely spits out some info messages about shutting down 🤷 

Think this is relevant? https://docs.python.org/3/library/logging.html#thread-safety

> If you are implementing asynchronous signal handlers using the [signal](https://docs.python.org/3/library/signal.html#module-signal) module, you may not be able to use logging from within such handlers. This is because lock implementations in the [threading](https://docs.python.org/3/library/threading.html#module-threading) module are not always re-entrant, and so cannot be invoked from such signal handlers.

I don't exactly know if by "asynchronous" here they mean `async def ...` or just "asynchronous" in general 🤔

**Comment by joerunde** [comment]
> Also, I don't see any unit tests

Yup! Didn't really have the bandwidth here and I couldn't find the gumption to try to patch in mock handlers that I could assert were called, but have them running in a subprocess...

File: tests/runtime/test_grpc_server.py [hunk 5]
```diff
@@ -1350,6 +1351,63 @@ def test_grpc_sever_shutdown_with_model_poll(open_port):
...
+            RUNTIME_METRICS_ENABLED="false",
+            RUNTIME_GRPC_ENABLED="true",
+            RUNTIME_HTTP_ENABLED="true",
+        )
+        with server_proc as proc:
+            # Wait for the grpc server to be up:
+            _assert_connection(
+                grpc.insecure_channel(f"localhost:{open_port}"), max_failures=500
+            )
+            print("grpc server is up!")
```
Comment (gabe-l-hart) [inline tests/runtime/test_grpc_server.py:1378]:
```suggestion
```

File: tests/runtime/test_grpc_server.py [hunk 6]
```diff
@@ -1350,6 +1351,63 @@ def test_grpc_sever_shutdown_with_model_poll(open_port):
...
+    """Test that a SIGINT successfully shuts down all running servers"""
+
+    # whoops, need 2 ports. Cannot use `_open_port()` again because it'll return the same port
+    other_open_port = open_port + 37
+
+    with tempfile.TemporaryDirectory() as workdir:
+        server_proc = ModuleSubproc(
+            "caikit.runtime",
+            kill_timeout=30.0,
+            RUNTIME_GRPC_PORT=str(open_port),
```
Comment (gabe-l-hart) [inline tests/runtime/test_grpc_server.py:1364]: Since this is testing with logs (meh), throw a `LOG_LEVEL` in there so that we guarantee the logs will be configured to print in case the parent invokes the test with logging turned all the way off

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
Two tiny nits!

File: tests/runtime/test_grpc_server.py (hunk 6 above)
Comment (joerunde) [reply tests/runtime/test_grpc_server.py:1364]: hot

File: tests/runtime/test_grpc_server.py (hunk 5 above)
Comment (joerunde) [reply tests/runtime/test_grpc_server.py:1378]: ahhhhh my debugs lol

File: tests/runtime/conftest.py [hunk 7]
```diff
@@ -69,10 +69,9 @@ def http_session_scoped_open_port():
     return _open_port()
 
 
-def _open_port():
+def _open_port(start=8888):
```
Comment (gabe-l-hart) [inline tests/runtime/conftest.py:72]: heh, we should probably move to [tls_test_tools.open_port()](https://github.com/IBM/tls-test-tools/blob/main/tls_test_tools/__init__.py#L39) for consistency, but certainly not now

**Review by gabe-l-hart — APPROVED** [review]
Looks good!
